src/backend/starknet.cairo

```diff
@@ -42,46 +43,54 @@ namespace Starknet {
         self: model.State*
     ) {
         alloc_locals;
-        let (native_token_address_) = native_token_address.read();
+        let (native_token_address) = Kakarot_native_token_address.read();
 
         // Accounts
         Internals._commit_accounts{state=self}(
-            self.accounts_start, self.accounts, native_token_address_
+            self.accounts_start, self.accounts, native_token_address
         );
 
         // Events
         Internals._emit_events(self.events_len, self.events);
 
         // Transfers
-        Internals._transfer_eth(native_token_address_, self.transfers_len, self.transfers);
+        Internals._transfer_eth(native_token_address, self.transfers_len, self.transfers);
 
         return ();
     }
 
-    // @notice Deploy a new account proxy
+    // @notice Deploy a new account
     // @dev Deploy an instance of an account
     // @param evm_address The Ethereum address which will be controlling the account
     // @param class_hash The hash of the implemented account (eoa/contract)
-    // @return account_address The Starknet Account Proxy address
+    // @return account_address The Starknet Account address
     func deploy{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
         class_hash: felt, evm_address: felt
     ) -> (account_address: felt) {
         alloc_locals;
+
+        // Deploy generic uninitialized account used to get a deterministic starknet address
         let (kakarot_address: felt) = get_contract_address();
-        let (_account_proxy_class_hash: felt) = account_proxy_class_hash.read();
+        let (
+            uninitialized_account_class_hash: felt
+        ) = Kakarot_uninitialized_account_class_hash.read();
         let (constructor_calldata: felt*) = alloc();
+        assert constructor_calldata[0] = kakarot_address;
+        assert constructor_calldata[1] = evm_address;
         let (starknet_address) = deploy_syscall(
-            _account_proxy_class_hash,
+            uninitialized_account_class_hash,
             contract_address_salt=evm_address,
-            constructor_calldata_size=0,
+            constructor_calldata_size=2,
             constructor_calldata=constructor_calldata,
             deploy_from_zero=0,
         );
-        assert constructor_calldata[0] = kakarot_address;
-        assert constructor_calldata[1] = evm_address;
-        IAccount.initialize(starknet_address, class_hash, 2, constructor_calldata);
+
+        // Properly initialize the account once created
+        let (account_class_hash) = Kakarot_generic_account_class_hash.read();
+        IUninitializedAccount.initialize(starknet_address, account_class_hash);
```

Comment:
I don't understant why we still need this two fold deployment. This was done originally to make sure that the deployed address doesn't depend on the type (EOA or CA). But now that they share the same class hash, we can avoid this.

Reply to self: we want to have an address independent of the account class hash to avoid future issue if we upgrade the account

Reply:
FYI: we use the same model in SSJ